test/hotspot/jtreg/compiler/lib/template_framework/Template.java

```diff
@@ -0,0 +1,599 @@
+                         BiFunction<A, B, TemplateBody> function) implements Template {
+        TemplateBody instantiate(A a, B b) {
+            return function.apply(a, b);
+        }
+
+        /**
+         * Creates a {@link TemplateWithArgs} which can be used as a {@link Token} inside
+         * a {@link Template} for nested code generation, and it can also be used with
+         * {@link TemplateWithArgs#render} to render the template to a {@link String}
+         * directly.
+         *
+         * @param a The value for the first argument.
+         * @param b The value for the second argument.
+         * @return The template all (two) arguments applied.
+         */
+        public TemplateWithArgs withArgs(A a, B b) {
+            return new TemplateWithArgs.TwoArgsUse<>(this, a, b);
+        }
+    }
+
+    /**
+     * A {@link Template} with three arguments.
+     *
+     * @param arg0Name The name of the first argument, used for hashtag replacements in the {@link Template}.
+     * @param arg1Name The name of the second argument, used for hashtag replacements in the {@link Template}.
+     * @param arg2Name The name of the third argument, used for hashtag replacements in the {@link Template}.
+     * @param <A> The type of the first argument.
+     * @param <B> The type of the second argument.
+     * @param <C> The type of the third argument.
+     * @param function The function with three arguments that creates the {@link TemplateBody} given the template arguments.
+     */
+    record ThreeArgs<A, B, C>(String arg0Name, String arg1Name, String arg2Name,
+                              TriFunction<A, B, C, TemplateBody> function) implements Template {
+        TemplateBody instantiate(A a, B b, C c) {
+            return function.apply(a, b, c);
+        }
+
+        /**
+         * Creates a {@link TemplateWithArgs} which can be used as a {@link Token} inside
+         * a {@link Template} for nested code generation, and it can also be used with
+         * {@link TemplateWithArgs#render} to render the template to a {@link String}
+         * directly.
+         *
+         * @param a The value for the first argument.
+         * @param b The value for the second argument.
+         * @param c The value for the third argument.
+         * @return The template all (three) arguments applied.
+         */
+        public TemplateWithArgs withArgs(A a, B b, C c) {
+            return new TemplateWithArgs.ThreeArgsUse<>(this, a, b, c);
+        }
+    }
+
+    /**
+     * Creates a {@link TemplateBody} from a list of tokens, which can be {@link String}s,
+     * boxed primitive types (e.g. {@link Integer}), any {@link Token}, or {@link List}s
+     * of any of these.
+     *
+     * {@snippet lang=java :
+     * var template = Template.make(() -> body(
+     *     """
+     *     Multi-line string
+     *     """,
+     *     "normal string ", Integer.valueOf(3), Float.valueOf(1.5f),
```

Comment:
Is `valueOf()` required or would it auto-box?

Reply:
Just updated the `TestTemplate.java`, and it would auto-box. I'll improve the example here.

Reply:
Expanded the example, and comments.